src/Symfony/Component/HttpKernel/ResettingRequestStack.php

```diff
@@ -0,0 +1,49 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpKernel;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+
+/**
+ * Resets services on second master requests.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class ResettingRequestStack extends RequestStack
```

Comment:
Could add an interface on the `RequestStack` and make this extension deprecated in 4.1 to promote composition over inheritance. Should probably also get `@final`